presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EvaluateProjectOnValues.java

```diff
@@ -0,0 +1,142 @@
+        List<List<RowExpression>> rows = source.getRows();
+        List<VariableReferenceExpression> valuesOutputVariables = source.getOutputVariables();
+        Set<Map.Entry<VariableReferenceExpression, RowExpression>> projectAssignmentEntries = projectNode.getAssignments().entrySet();
+        List<VariableReferenceExpression> projectOutputVariables = projectAssignmentEntries.stream().map(Map.Entry::getKey).collect(toImmutableList());
+        List<RowExpression> projectRowExpressions = projectAssignmentEntries.stream().map(Map.Entry::getValue).collect(toImmutableList());
+        // exclude non-deterministic function
+        if (!projectRowExpressions.stream()
+                .filter(expression -> expression instanceof CallExpression)
+                .map(callExpression -> functionAndTypeManager.getFunctionMetadata(((CallExpression) callExpression).getFunctionHandle()))
+                .filter(functionMetadata -> !functionMetadata.isDeterministic()).collect(toImmutableList()).isEmpty()) {
+            return Result.empty();
+        }
+
+        //evaluate ProjectNode
+        List<List<Object>> rowValues = rows.stream().map(row -> {
+            //Prepare to set up variable resolver
+            verify(row.size() == valuesOutputVariables.size(), "Output variable does not match its value in ValuesNode");
+            Map<String, Object> valuesMapForResolver = Streams.zip(valuesOutputVariables.stream(),
+                    row.stream(),
+                    (valuesOutputVariable, element) -> {
+                        RowExpressionInterpreter elementInterpreter = new RowExpressionInterpreter(element,
+                                functionAndTypeManager,
+                                context.getSession().toConnectorSession(),
+                                EVALUATED);
+                        return new AbstractMap.SimpleImmutableEntry<String, Object>(valuesOutputVariable.getName(), elementInterpreter.evaluate());
+                    }).collect(toImmutableMap(Map.Entry::getKey, Map.Entry::getValue));
+            VariableResolver variableResolver = new Interpreters.LambdaVariableResolver(valuesMapForResolver);
+            //evaluate each row of the ProjectNode
+            return projectRowExpressions.stream().map(rowExpression -> new RowExpressionInterpreter(
+                            rowExpression,
+                            functionAndTypeManager,
+                            context.getSession().toConnectorSession(),
+                            OPTIMIZED).optimize(variableResolver))
+                    .collect(toImmutableList());
+        }).collect(toImmutableList());
+
+        //Form the ValuesNode transformed from ProjectNode
+        List<List<RowExpression>> valuesRows = rowValues.stream().map(row ->
+                        Streams.zip(row.stream(), projectOutputVariables.stream(), (elementValue, projectOutputVariable) -> (RowExpression) new ConstantExpression(elementValue, projectOutputVariable.getType()))
```

Comment:
This line is very long. Can you put the arguments on new lines?

Reply:
sure